library/src/G4LatticeLogical.cc

```diff
@@ -690,6 +691,109 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
   fValleyAxis.push_back(fValleyInv.back()*G4ThreeVector(1.,0.,0.));
 }
 
+// Store drifting-electron valley using valley's direction
+
+void G4LatticeLogical::AddValley(const G4ThreeVector& valleyDirVec, G4bool antival) {
+      
+  // Find the rotation matrix elements. 
+  // The rows correspond to the rotated axis direction 
+  // First row is vx,vy,vz ; second row is a,b,c and third row is d,e,f
+    
+  // We chose the following convention :
+  //  - The rotated y axis must stay in the X-Y plane (c=0)
+  //  - Valley's rotated Z axis must have positive z component (f>0) 
+  //  - Anti-valley's rotated Z axis must have positive z component (f>0) 
+  //  - The three rotated axis should stay normalized :
+  //	vx^2 + vy^2 + vz^2 = 1 ; a^2 + b^2 = 1 ; d^2 + e^2 + f^2 = 1
+  //  - And be orthogonal :
+  //	ad + be = 0 ; avx + bvy = 0 ; dvx + evy + fvz = 0
+  //  - It must be a right hand coordinate system :
+  //	(vx,vy,vz) X (a,b,c) = (d,e,f) : d=-vzb ; e=vza ; f=vxb - vya
+    
+  // We first compute f 
+  // d^2 + e^2 + f^2 = 1
+  // f^2 = 1 - d^2 - e^2
+  // f^2 = 1 - vz^2 b^2 - vz^2 a^2
+  // f^2 = 1 - vz^2 (b^2 + a^2)
+  // f^2 = 1 - vz^2
+  // f = +- sqrt(1 - vz^2)
+
+  // We compute b (if vx=!0 and vy!=0) : 
+  // f = vxb - vya
+  // f = vxb - vy (-bvy/vx)
+  // f = b/vx(vx^2 + vy^2)
+  // b = fvx/(vx^2 + vy^2)
+  // b = fvx/(1 - vz^2)
+  // b = vx/f
+    
+  // Then we compute a (if vx=!0 and vx!=0) : 
+  // a = -bvy/vx
+  // a = -vxvy/e/vx
+  // a = -vy/e
+
+  // Finally, we compute d and e :
+  // d=-vzb
+  // e=vza 
+    
+  // If vx=0 and/or vy=0, we compute a and b differently.
+
+  tempvec()=valleyDirVec.unit();
+    
+  double vx=tempvec().x();
+  double vy=tempvec().y();
+  double vz=tempvec().z();
+  
+  double a;
+  double b;
+  double c=0; // rotated y axis must stay in the X-Y plane
+  double d;
+  double e;
+  double f;
```

Comment:
Don't need to declare these in advance.  Just declare the variable in the one-line calculation.